skyportal/api_description.md

```diff
@@ -110,3 +110,50 @@ for the failure given in `message`:
   "version": "0.9.1"
 }
 ```
+
+### Pagination
+
+Several API endpoints (notably the sources and candidates APIs) enforce
+pagination to limit the number of records that can be fetched per request.
+These APIs expose parameters that facilitate pagination (see the various
+API docs for details). A sample pagination script is included here:
+
+```python
+import requests
+import time
+
+
+base_url = "https://fritz.science/api/sources"
+token = "your_token_id_here"
+group_id = 4  # If applicable
+all_sources = []
+num_per_page = 500
+page = 1
+total_matches = None
+retry_attempts = 0
+max_retry_attempts = 10
+
+while retry_attempts <= max_retry_attempts:
+    r = requests.get(
+        f"{base_url}?group_ids={group_id}&pageNumber={page}&numPerPage={num_per_page}&totalMatches={total_matches}",
+        headers={"Authorization": f"token {token}"},
+    )
+    data = r.json()
+    if data["status"] != "success":
+        print(data)  # log as appropriate
+        retry_attempts += 1
+        time.sleep(1)
```

Comment:
This is arbitrary and is not required on our end.

Reply:
So this is a case in which the server is likely unavailable due to error.

Ari, should we have delays so that the script can never be rejected to our burst limits?  That way, when an error occurs, we know we need to wait a while (1s probably inadequate) since the server is busy/down.

Reply:
Our rate limiting is based on requests per second, so waiting 1 second should be fine here right?

Reply:
I think it is unlikely that this script will ever hit the rate limit, but we can ensure that by delaying the requests that go out—by 0.2s, to avoid the 5r/s limit.

Then, we only have to worry about other error codes, in which case 1s would be fine, but is unlikely to help—so may be worth backing off 5s instead.  Server recovery is on the order of 30s, so 10 attempts at a 5s delay should get them there.